what4/src/What4/Solver/Boolector.hs

```diff
@@ -58,7 +58,7 @@ boolectorOptions =
       executablePathOptSty
       (Just "Path to boolector executable")
       (Just (ConcreteString "boolector"))
-  ]
+  ] <> SMT2.smtlib2Options
```

Comment:
I don't think we can just add the smtlib2Options to each individual solver like this.  If we install the options for more than one SMTLib2 solver at a time, I think the option will clash and we'll get an error from the `Config` module.

I'm not sure what the best way to deal with this is.  Maybe we should just silently ignore adding options that already exist in the map, if they have the right types.

Reply:
OK, this ended up with an overhaul of the configs to allow idempotency and deprecation, and it switched the solver configs to be heirarchical so that it should be easy to have per-solver overrides when we want them.

This made it a big pull request; normally I would keep them smaller and I could split this one up, but it will take a little work to detangle the different sets of changes, so let me know what you would like.